File: zabbix_get_jmx.py
#!/usr/bin/env python3
# vim: set ai ts=2 sw=2 expandtab :
# Author: Nour Sharabash <nour.sharabash@gmail.com>

import argparse, sys, json, re, socket

args = None
def parse_args():
  global args
  parser = argparse.ArgumentParser()
  parser.add_argument("--java-gateway-host")
  parser.add_argument("--java-gateway-port")
  parser.add_argument("--jmx-server")
  parser.add_argument("--jmx-port")
  parser.add_argument("--key")
  parser.add_argument("--jmx-user")
  parser.add_argument("--jmx-pass")
  parser.add_argument("--new-protocol", default=False, action='store_true')
  args = parser.parse_args(sys.argv[1:])

def main():
  parse_args()
  query = { 'request': 'java gateway jmx',
             'conn': args.jmx_server, 'port': int(args.jmx_port),
             'keys': [args.key] }
  if args.jmx_user and args.jmx_pass:
    query['username'] = args.jmx_user
    query['password'] = args.jmx_pass
  if args.new_protocol:
    conn = query.pop('conn')
    port = query.pop('port')
    query['jmx_endpoint'] = 'service:jmx:rmi:///jndi/rmi://%s:%s/jmxrmi' % (conn, port)


  query_str = json.dumps(query)
  query_len = len(query_str)
  query_len_hex = '%.16x' % query_len
  query_len_bin = re.sub(r'^(..)(..)(..)(..)(..)(..)(..)(..)', '\\x\\8\\x\\7\\x\\6\\x\\5\\x\\4\\x\\3\\x\\2\\x\\1', query_len_hex)
  query_bin = "ZBXD\\x01%s" % (query_len_bin)
  query_bin = query_bin.encode('latin-1').decode('unicode_escape')
  query_bin += query_str

  with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
    s.connect((args.java_gateway_host, int(args.java_gateway_port)))
    s.send(bytes(query_bin, 'latin-1'))
    size = 1024
    data = s.recv(size)
    full = ''
    started = False
    while len(data):
      data = str(data, 'latin-1')
      if not started:
        if '{' in data:
          data = data[data.index('{'):]
          started = True
        else:
          data = ''
      if started:
        full += data
      data = s.recv(size)
    print(full)

if __name__ == '__main__': main()

File: zabbix_get_jmx.rb
#!/usr/bin/ruby
# vim: set ai ts=2 sw=2 expandtab ft=ruby syn=ruby :
# Author: Nour Sharabash <nour.sharabash@gmail.com>
require 'json'
require 'socket'
require 'getoptlong'

args = GetoptLong.new(
  [ '--java-gateway-host', GetoptLong::OPTIONAL_ARGUMENT ],
  [ '--java-gateway-port', GetoptLong::OPTIONAL_ARGUMENT ],
  [ '--jmx-server', GetoptLong::OPTIONAL_ARGUMENT ],
  [ '--jmx-port', GetoptLong::OPTIONAL_ARGUMENT ],
  [ '--key', GetoptLong::OPTIONAL_ARGUMENT ],
  [ '--jmx-user', GetoptLong::OPTIONAL_ARGUMENT ],
  [ '--jmx-pass', GetoptLong::OPTIONAL_ARGUMENT ],
  [ '--new-protocol', GetoptLong::OPTIONAL_ARGUMENT ]
)

query = {'request': 'java gateway jmx'}
gateway = {}
args.each do |opt, arg|
  case opt
    when '--java-gateway-host'
      gateway['host'] = arg
    when '--java-gateway-port'
      gateway['port'] = arg.to_i
    when '--jmx-server'
      query['conn'] = arg
    when '--jmx-port'
      query['port'] = arg.to_i
    when '--key'
      query['keys'] = [arg]
    when '--jmx-user'
      query['username'] = arg
    when '--jmx-pass'
      query['password'] = arg
    when '--new-protocol'
      conn = query.delete 'conn'
      port = query.delete 'port'
      query['jmx_endpoint'] = "service:jmx:rmi:///jndi/rmi://#{conn}:#{port}/jmxrmi"
  end
end
query_str = JSON.dump(query)
query_len = query_str.length
query_len_hex = '%.16x' % query_len
query_len_bin = query_len_hex.gsub(/^(..)(..)(..)(..)(..)(..)(..)(..)/, '\\x\8\\x\7\\x\6\\x\5\\x\4\\x\3\\x\2\\x\1')
query_bin = eval('"'+ "ZBXD\\x01#{query_len_bin}" '"').bytes.pack('C*') + query_str

s = TCPSocket.new(gateway['host'], gateway['port'])
s.write query_bin
size = 1024
data = s.read(size)
full = ''
started = false
while data and data.length
  if not started
    if data.index('{')
      data = data[data.index('{'), data.length]
      started = true
    else
      data = ''
    end
  end
  if started
    full += data
  end
  data = s.read(size)
end
s.close
puts full
